### PR TITLE
Run security rules from PSRule by default, and exclude repeated rules

### PR DESCRIPTION
**Pull request by VeraBE**
## Description

Added a baseline to run the PSRule rules under the [Security pillar](https://azure.github.io/PSRule.Rules.Azure/en/rules/module/#security) by default, except the ones already present in TemplateAnalyzer; and another one to exclude those repeated rules when using `--run-all-rules`

Documentation on PSRule baselines: https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Baseline/

---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [x] The pull request does not introduce breaking changes.

### [General Guidelines](../CONTRIBUTING.md)
- [x] Title of the pull request is clear and informative.
- [x] Description of the pull request is clear and informative.
- [x] I have added myself to the 'assignees'.
- [x] I have added 'linked issues' if relevant.

### [Testing Guidelines](../CONTRIBUTING.md#tests)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -35,7 +35,7 @@ Argument | Description
 **(Optional)** `--report-format` | Valid formats:<br/>*Console*: output results to the console in plain text. **(default)**<br/>*Sarif*: output results to a file in [SARIF](https://sarifweb.azurewebsites.net) format.
 `-o` or `--output-file-path` | **(Required if `--report-format` is *Sarif*)**  File path to output SARIF results to.
 **(Optional)** `-v` or `--verbose` | Shows details about the analysis
-**(Optional)** `--run-powershell` | Also runs the PowerShell-based rules
+**(Optional)** `--run-all-rules` | Run all the rules against the templates, including non-security ones
```
Comment (JohnathonMohr) [inline README.md:38]: How about '--include-non-security-rules' instead?  On first glance it seems odd, and I wonder "wait, why do I have to manually specify I want to run them all?"  Having "non-security" specified up-front clarifies it.

Comment (JohnathonMohr) [reply README.md:38]: Also, this flag only applies to PSRule at the moment.  This might get complex if we introduce non-security checks in JSON rules as well.

File: docs/customizing-evaluation-outputs.md [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # Customizing Evaluation Outputs
 
 ## Overview
-Template BPA customization is authored in JSON.  Custom evaluation settings are written in a separate configuration and used on tool execution. See the main README for details on [how to run Template BPA with a configuration file](https://github.com/Azure/template-analyzer#using-the-template-bpa). 
+Template BPA customization is authored in JSON. Custom evaluation settings are written in a separate configuration and used on tool execution. See the main README for details on [how to run Template BPA with a configuration file](https://github.com/Azure/template-analyzer#using-the-template-bpa). It currently only supports the JSON-based rules.
```
Comment (JohnathonMohr) [inline docs/customizing-evaluation-outputs.md:4]:
```suggestion
Using an additional JSON configuration file, the Template BPA can be customized in how it runs JSON-based rules.  For example, specific rules can be included in/excluded from execution, or the severity of a rule can be changed. See the main README for details on [how to run Template BPA with a configuration file](https://github.com/Azure/template-analyzer#using-the-template-bpa).

**Note:** The configuration file only affects JSON-based rules; PSRule rules cannot be configured here at this time.
```

File: src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs [hunk 3]
```diff
@@ -115,26 +104,14 @@ public void AnalyzeTemplate_BicepTemplateIsInvalid_ThrowTemplateAnalyzerExceptio
...
             }
             finally
             {
                 File.Delete(templateFilePath);
             }
         }
 
-        [TestMethod]
-        [ExpectedException(typeof(TemplateAnalyzerException))]
-        public void AnalyzeTemplate_MissingFilePathWithPowerShellOn_ThrowTemplateAnalyzerException()
```
Comment (JohnathonMohr) [inline src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs:128]: Why was this test removed?

File: src/Analyzer.PowerShellRuleEngine.UnitTests/Analyzer.PowerShellRuleEngine.UnitTests.csproj [hunk 4]
```diff
@@ -24,4 +24,8 @@
     </Content>
   </ItemGroup>
 
+  <ItemGroup>
+    <None Remove="templates\excluded_rules.json" />
```
Comment (JohnathonMohr) [inline src/Analyzer.PowerShellRuleEngine.UnitTests/Analyzer.PowerShellRuleEngine.UnitTests.csproj:28]: Was this meant to be deleted?  It's still added in the PR.

File: src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs [hunk 5]
```diff
@@ -69,32 +82,97 @@ public void AnalyzeTemplate_ValidTemplate_ReturnsExpectedEvaluations(string temp
...
+            else
+            {
+                var powerShellRuleEngineSecurityRules = new PowerShellRuleEngine(false);
+                evaluations = powerShellRuleEngineSecurityRules.AnalyzeTemplate(templateContext);
+            }
+
+            Assert.IsTrue(!evaluations.Any(evaluation => evaluation.RuleId == "AZR-000081")); // It's the id of Azure.AppService.RemoteDebug
+
+            // The RepeatedRulesBaseline will only be used when all rules are run:
+            if (runsAllRules)
```
Comment (JohnathonMohr) [inline src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs:127]: Should this be removed?  Otherwise line 143 will never be hit.

File: src/Analyzer.PowerShellRuleEngine.UnitTests/Analyzer.PowerShellRuleEngine.UnitTests.csproj (hunk 4 above)
Comment (JohnathonMohr) [reply src/Analyzer.PowerShellRuleEngine.UnitTests/Analyzer.PowerShellRuleEngine.UnitTests.csproj:28]: Ah, it's used in a test.  Please remove this change.

File: src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs [hunk 6]
```diff
@@ -69,32 +82,97 @@ public void AnalyzeTemplate_ValidTemplate_ReturnsExpectedEvaluations(string temp
...
+                var matchingEvaluation = failedEvaluations.Find(evaluation => evaluation.Result.LineNumber == expectedLineNumber);
+                failedEvaluations.Remove(matchingEvaluation);
+            }
+            Assert.IsTrue(failedEvaluations.IsEmptyEnumerable());
+        }
+
+        [DataTestMethod]
+        [DataRow(true, DisplayName = "Repeated rules are excluded when running all the rules")]
+        [DataRow(true, DisplayName = "Repeated rules are excluded when running only the security rules")]
+        public void AnalyzeTemplate_ValidTemplate_ExcludesRepeatedRules(bool runsAllRules)
```
Comment (JohnathonMohr) [inline src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs:97]: I'm not sure what's supposed to be tested here.  The title describes repeated rules being excluded, but the only assertion in the test is whether or not a specific rule appears, which is related to whether or not non-security rules are run, unaffected by repetition.

File: src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs [hunk 7]
```diff
@@ -74,13 +82,37 @@ public IEnumerable<IEvaluation> AnalyzeTemplate(TemplateContext templateContext)
...
+                        Outcome = RuleOutcome.Fail,
+                        Culture = new string[] { "en-US" } // To avoid warning messages when running tests in Linux
+                    },
+                    Include = new IncludeOption
+                    {
+                        Path = new string[]
+                        {
+                            ".ps-rule",
+                            Path.Combine(Path.GetDirectoryName(AppContext.BaseDirectory), "baselines", "SecurityBaseline.Rule.json"),
+                            ".ps-rule",
```
Comment (JohnathonMohr) [inline src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs:94]: Duplicate?

File: src/Analyzer.Reports.UnitTests/SarifReportWriterE2ETests.cs [hunk 8]
```diff
@@ -40,32 +40,49 @@ public void AnalyzeTemplateTests()
             }
 
             // assert
```
Comment (JohnathonMohr) [inline src/Analyzer.Reports.UnitTests/SarifReportWriterE2ETests.cs:42]: After these changes, the E2E tests no longer validate the scenario where a single rule has multiple distinct failures.  Previously, TA-000028 had two failing results, each with 3 failing lines.

File: src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs (hunk 3 above)
Comment (VeraBE) [reply src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs:128]: It was there to validate a check that doesn't exist anymore, the file path now is a required parameter of the function, it will always fail if it's not specified

File: src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs (hunk 6 above)
Comment (VeraBE) [reply src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs:97]: There are two assertions in the code. One checks that whether running all the rules or not an excluded rule is not executed, and the second one validates that when the exclusion baseline is not defined properly the rule is indeed triggering with the test file

File: README.md (hunk 1 above)
Comment (VeraBE) [reply README.md:38]: Thoughts from someone else before I refactor the name everywhere?

File: src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs (hunk 3 above)
Comment (JohnathonMohr) [reply src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs:128]: > it will always fail if it's not specified

I would argue that behavior should be validated with this test.

File: src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs (hunk 6 above)
Comment (JohnathonMohr) [reply src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs:97]: I see.  Can you add a comments explaining more of how the test works? It would be worth mentioning that the specific id you're looking for is a duplicate of a JSON rule, and should be excluded, unless the baseline file is missing (moved in the test).

File: src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs (hunk 3 above)
Comment (VeraBE) [reply src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs:128]: Do we validate for all our functions that they fail when we don't pass a required argument? :P

Comment (JohnathonMohr) [reply src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs:128]: We should.  If it's required, we should validate that the code enforces the requirement.  Example tests for LeafExpression: https://github.com/Azure/template-analyzer/blob/development/src/Analyzer.JsonRuleEngine.UnitTests/LeafExpressionTests.cs#L144#L156

Comment (JohnathonMohr) [reply src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs:128]: Put another way, if it's required, but there's a logical path in code where that requirement is somehow bypassed, there's a bug.  So the requirement should be tested.

Comment (VeraBE) [reply src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs:128]: I think we are not understanding each other, that test doesn't even compile anymore, because the function call is missing a parameter

Comment (VeraBE) [reply src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs:128]: I think we are not understanding each other, that test doesn't even compile anymore, because the function call is missing a parameter. It doesn't make sense to test what happens when the parameter is missing, because it doesn't compile when that happens

Comment (JohnathonMohr) [reply src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs:128]: Previously it was set to null by default, so it could be "left out".  The test should pass `null` as the value and expect a failure.

Comment (JohnathonMohr) [reply src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs:128]: Maybe "missing" vs. "null" is the confusing piece.  It should validate a failure from passing null as the file path.

File: README.md (hunk 1 above)
Comment (BernieWhite) [reply README.md:38]: @JohnathonMohr @VeraBE Alternative suggestion:

`--baseline` or `--with-baseline`

PSRule has many baselines. By default, customers get the latest security baseline for BPA or they can specify one. 

For example:

`--baseline Azure.All` or `--baseline Azure.Preview`

**Review by BernieWhite — APPROVED** [review]
@VeraBE Just one additional comments otherwise all good.

File: src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs [hunk 9]
```diff
@@ -42,7 +47,15 @@ public void AnalyzeTemplate_ValidTemplate_ReturnsExpectedEvaluations(string temp
...
             };
 
-            var evaluations = powerShellRuleEngine.AnalyzeTemplate(templateContext);
+            IEnumerable<IEvaluation> evaluations;
+            if (runsAllRules)
+            {
+                evaluations = powerShellRuleEngineAllRules.AnalyzeTemplate(templateContext);
+            } else
+            {
+                var powerShellRuleEngineSecurityRules = new PowerShellRuleEngine(false);
```
Comment (anaismiller) [inline src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs:56]: Why don't you put this at the top too?

File: src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs [hunk 10]
```diff
@@ -15,19 +17,22 @@ namespace Microsoft.Azure.Templates.Analyzer.RuleEngines.PowerShellEngine.UnitTe
...
         [AssemblyInitialize]
         public static void AssemblyInitialize(TestContext context)
         {
-            powerShellRuleEngine = new PowerShellRuleEngine();
+            powerShellRuleEngineAllRules = new PowerShellRuleEngine(true);
         }
 
         [DataTestMethod]
-        [DataRow("template_and_resource_level_results.json", 12, new int[] { 1, 1, 1, 1, 8, 14, 17, 1, 17, 17, 1, 17 }, DisplayName = "Template with errors reported in both analysis stages")]
+        [DataRow("template_and_resource_level_results.json", true, 12, new int[] { 1, 1, 1, 1, 8, 14, 17, 1, 17, 17, 1, 17 }, DisplayName = "Running all the rules against a template with errors reported in both analysis stages")]
```
Comment (anaismiller) [inline src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs:29]: What does `in both analysis stages` mean?

Comment (anaismiller) [reply src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs:29]: oh tool errors and rule violations?

File: src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs [hunk 11]
```diff
@@ -69,32 +82,89 @@ public void AnalyzeTemplate_ValidTemplate_ReturnsExpectedEvaluations(string temp
...
+            }
+            Assert.IsTrue(failedEvaluations.IsEmptyEnumerable());
+        }
+
+        [DataTestMethod]
+        [DataRow(true, DisplayName = "Repeated rules are excluded when running all the rules")]
+        [DataRow(true, DisplayName = "Repeated rules are excluded when running only the security rules")]
+        public void AnalyzeTemplate_ValidTemplate_ExcludesRepeatedRules(bool runsAllRules)
+        {
+            var templateFilePath = Path.Combine(templatesFolder, "excluded_rules.json");
```
Comment (anaismiller) [inline src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs:99]: Can you rename the file to TestExcludedRules.json?

Comment (anaismiller) [reply src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs:99]: I thought it was a rule file or configuration file and it confused me.

File: src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs [hunk 12]
```diff
@@ -243,10 +243,18 @@ public void FilterRules_ValidConfig_RulesFiltered(bool isBicep, string configNam
...
                 if (specifyInCommand)
+                {
                     args = args.Concat(new[] { "--config-file-path", configName }).ToArray();
+                }
+
+                using StringWriter outputWriter = new();
+                Console.SetOut(outputWriter);
 
                 result = _commandLineParser.InvokeCommandLineAPIAsync(args);
-                Assert.AreEqual((int)ExitCode.Success, result.Result);
```
[the remark this answers is not in the document]
Comment (VeraBE) [reply src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs:249]: It doesn't work anymore because all those files return PSRule violations :/

File: src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs (hunk 10 above)
Comment (VeraBE) [reply src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs:29]: No, PSRule detects errors when looking at the whole file (through the file path), and when looking at each resource (`pipeline.Process(resource);`)

File: src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs (hunk 11 above)
Comment (VeraBE) [reply src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs:99]: Do you think it's necessary even considering it is in the "templates" folder?

File: src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (anaismiller) [reply src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs:249]: It does not have to be ExitCode.Success. And the value can be passed in as a parameter in the test function if each one is different

File: src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs (hunk 10 above)
Comment (anaismiller) [reply src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs:29]: Oh! Maybe this is something that should be added to the ReadMe? Not sure.

File: src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs (hunk 11 above)
Comment (anaismiller) [reply src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs:99]: If I plan to make a change to this test, it could help me save time understanding what it going on. And if I want to use this file in a test, would also help. 
If there is a different name, could also be good.

File: README.md (hunk 1 above)
Comment (VeraBE) [reply README.md]: I'm not sure if I like the idea of having a config file to filter the JSON rules and the baseline file for the PSRule rules, shouldn't we try to keep a consistent interface for all types of rules in the tool? Maybe by translating the config file into baseline files? Even though it sounds a bit annoying because in our config file we use rule ids and the baseline needs rule names

Comment (JohnathonMohr) [reply README.md]: I agree, we *should* keep a consistent interface for configuring all rules, regardless of where/how they're written.  That's going to require a thoughtful design though, and shouldn't block our current progress.  For now, I think ignoring the available baselines and just keeping what we have (running just security rules by default, but allowing a way to additionally run non-security rules if desired) is best; then we can figure out how to continue the integration next.

**Review by anaismiller — APPROVED** [review]

File: src/Analyzer.Core/TemplateAnalyzer.cs [hunk 13]
```diff
@@ -47,10 +47,10 @@ private TemplateAnalyzer(JsonRuleEngine jsonRuleEngine, PowerShellRuleEngine pow
         /// <summary>
         /// Creates a new <see cref="TemplateAnalyzer"/> instance with the default built-in rules.
         /// </summary>
-        /// <param name="usePowerShell">Whether or not to use PowerShell rules to analyze the template.</param>
+        /// <param name="runAllRules">Whether or not to run all the rules against the template.</param>
         /// <param name="logger">A logger to report errors and debug information</param>
         /// <returns>A new <see cref="TemplateAnalyzer"/> instance.</returns>
-        public static TemplateAnalyzer Create(bool usePowerShell, ILogger logger = null)
+        public static TemplateAnalyzer Create(bool runAllRules, ILogger logger = null)
```
Comment (JohnathonMohr) [inline src/Analyzer.Core/TemplateAnalyzer.cs:53]: Rename here as well.

File: src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs [hunk 14]
```diff
@@ -27,9 +33,11 @@ public class PowerShellRuleEngine : IRuleEngine
         /// <summary>
         /// Creates a new instance of a PowerShellRuleEngine.
         /// </summary>
+        /// <param name="runAllRules">Whether or not to run all the rules against the template.</param>
         /// <param name="logger">A logger to report errors and debug information.</param>
-        public PowerShellRuleEngine(ILogger logger = null)
+        public PowerShellRuleEngine(bool runAllRules, ILogger logger = null)
```
Comment (JohnathonMohr) [inline src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs:38]: Rename here and class variable too.

File: src/Analyzer.PowerShellRuleEngine/Analyzer.PowerShellRuleEngine.csproj [hunk 15]
```diff
@@ -8,13 +8,19 @@
   
   <ItemGroup>
     <PackageReference Include="Microsoft.PowerShell.SDK" Version="7.2.4" />
-    <PackageReference Include="Microsoft.PSRule.Rules.Azure" Version="1.18.0-B0010" />
-    <PackageReference Include="Microsoft.PSRule.SDK" Version="2.3.0-B0130" />
+    <PackageReference Include="Microsoft.PSRule.Rules.Azure" Version="1.19.0-B0010" />
```
Comment (JohnathonMohr) [inline src/Analyzer.PowerShellRuleEngine/Analyzer.PowerShellRuleEngine.csproj:11]: Please update the nuspec file for Analyzer.Core too.  We eventually need a better way to maintain that (created tracking issue #279), but for now, we need to update manually.

**Review by JohnathonMohr — CHANGES_REQUESTED** [review]

File: src/Analyzer.PowerShellRuleEngine/Analyzer.PowerShellRuleEngine.csproj (hunk 15 above)
Comment (JohnathonMohr) [reply src/Analyzer.PowerShellRuleEngine/Analyzer.PowerShellRuleEngine.csproj:11]: There are also other out-of-date versions in that nuspec, as well as dependencies that are missing entirely (Bicep).  Can you update those as well?

**Review by JohnathonMohr — APPROVED** [review]
